### ISSUE TITLE
FTWRL way too long using mydumper 0.12.3-3

### ISSUE THREAD

**Issue by vnlo77250** [mydumper/mydumper#751]
A comparison between mydumper 0.9.1 and mydumper show that mydumper 0.9.1 release lock way faster (in about 50 minutes).

- mydumper 0.9.1 dump non innodb table before innodb table.
- with mydumper 0.12.3, after a hour, not a single non innodb table was dumped, while table schema and data of innodb table were already dumped.

In file mydumper_working_thread.c, i can see the following lines should not work for MYSQL, as engine MRG_MYISAM doesn't show on SHOW TABLE STATUS, but MYISAM is shown instead.

    if (!no_data) {
      if (ecol != NULL && g_ascii_strcasecmp("MRG_MYISAM",ecol)) {
        if (data_checksums) {
          create_job_to_dump_checksum(dbt, conf);
        }
        if (trx_consistency_only ||
          (ecol != NULL && (!g_ascii_strcasecmp("InnoDB", ecol) || !g_ascii_strcasecmp("TokuDB", ecol)))) {
          create_job_to_dump_table(conn, dbt, conf, TRUE);
        } else {
          g_mutex_lock(non_innodb_table_mutex);
          non_innodb_table = g_list_prepend(non_innodb_table, dbt);
          g_mutex_unlock(non_innodb_table_mutex);
        }
      }
    }

** 

**To Reproduce**
Command executed:
* /usr/bin/mydumper --user="db-backup" --password="*******" --host="10.93.100.1" --port="3306" --outputdir /mnt/backup/O2S-BDD11/20220710202251 -t 16 -E  --compress --routines --triggers --verbose=3

**Expected behavior**
        Dump of non innodb table shoud have been prioritize, and unlock tables use before dumping innodb table

**Log**
Log of mydumper 0.9.1
2022-07-10 22:00:02 [INFO] - Connected to a MySQL server
2022-07-10 22:00:11 [INFO] - Started dump at: 2022-07-10 22:00:11

2022-07-10 22:00:11 [INFO] - Written master status
2022-07-10 22:00:11 [INFO] - Written slave status
2022-07-10 22:00:11 [INFO] - Thread 1 connected using MySQL connection ID 299477
2022-07-10 22:00:11 [INFO] - Thread 2 connected using MySQL connection ID 299479
2022-07-10 22:00:11 [INFO] - Thread 3 connected using MySQL connection ID 299481
2022-07-10 22:00:11 [INFO] - Thread 4 connected using MySQL connection ID 299483
2022-07-10 22:00:11 [INFO] - Thread 5 connected using MySQL connection ID 299484
2022-07-10 22:00:11 [INFO] - Thread 6 connected using MySQL connection ID 299485
2022-07-10 22:00:11 [INFO] - Thread 7 connected using MySQL connection ID 299486
2022-07-10 22:00:11 [INFO] - Thread 8 connected using MySQL connection ID 299487
2022-07-10 22:00:11 [INFO] - Thread 9 connected using MySQL connection ID 299488
2022-07-10 22:00:11 [INFO] - Thread 10 connected using MySQL connection ID 299489
2022-07-10 22:00:11 [INFO] - Thread 11 connected using MySQL connection ID 299490
2022-07-10 22:00:11 [INFO] - Thread 12 connected using MySQL connection ID 299491
2022-07-10 22:00:11 [INFO] - Thread 13 connected using MySQL connection ID 299492
2022-07-10 22:00:11 [INFO] - Thread 14 connected using MySQL connection ID 299493
2022-07-10 22:00:11 [INFO] - Thread 15 connected using MySQL connection ID 299494
2022-07-10 22:00:12 [INFO] - Thread 16 connected using MySQL connection ID 299495
2022-07-10 22:50:13 [INFO] - Queuing Non-InnoDB tables
2022-07-10 22:50:13 [INFO] - Thread 2 dumping data for `mysql`.`columns_priv`
2022-07-10 22:50:13 [INFO] - Thread 1 dumping data for `mysql`.`db`
2022-07-10 22:50:13 [INFO] - Thread 3 dumping data for `mysql`.`event`
2022-07-10 22:50:13 [INFO] - Thread 4 dumping data for `mysql`.`func`
2022-07-10 22:50:13 [INFO] - Thread 6 dumping data for `mysql`.`host`
2022-07-10 22:50:13 [INFO] - Thread 5 dumping data for `mysql`.`ndb_binlog_index`
2022-07-10 22:50:13 [INFO] - Thread 9 dumping data for `mysql`.`proc`
2022-07-10 22:50:13 [INFO] - Thread 10 dumping data for `mysql`.`tables_priv`
2022-07-10 22:50:13 [INFO] - Thread 7 dumping data for `mysql`.`procs_priv`
2022-07-10 22:50:13 [INFO] - Thread 12 dumping data for `mysql`.`user`
2022-07-10 22:50:13 [INFO] - Thread 13 dumping data for `mysql`.`user_info`
2022-07-10 22:50:13 [INFO] - Thread 11 dumping data for `victoriacapital`.`port_flux_copy`
2022-07-10 22:50:13 [INFO] - Thread 14 dumping data for `victoriacapital`.`port_situations_copy`
2022-07-10 22:50:13 [INFO] - Empty table mysql.columns_priv
2022-07-10 22:50:13 [INFO] - Empty table mysql.host
2022-07-10 22:50:13 [INFO] - Empty table mysql.func
2022-07-10 22:50:13 [INFO] - Empty table mysql.ndb_binlog_index
2022-07-10 22:50:13 [INFO] - Empty table mysql.procs_priv
2022-07-10 22:50:13 [INFO] - Empty table mysql.user_info
2022-07-10 22:50:55 [INFO] - Non-InnoDB dump complete, unlocking tables
2022-07-10 22:50:55 [INFO] - Queuing InnoDB tables
2022-07-10 22:50:55 [INFO] - Thread 16 dumping data for `1001patrimoines`.`actions`
2022-07-10 22:50:55 [INFO] - Thread 15 dumping data for `1001patrimoines`.`actions_groupes`

Extract of Dump of mydumper 0.12.3 (only database mysql possess non innodb tables), we can see non queuing of mysql table.
2022-07-10 20:50:33 [INFO] - Thread 12 dumping db information for `yarokcapital`
2022-07-10 20:50:34 [INFO] - Thread 9 dumping db information for `ydfinance`
2022-07-10 20:50:34 [INFO] - Thread 3 dumping db information for `yetcpatrimoine`
2022-07-10 20:50:35 [INFO] - Thread 15 dumping db information for `ypatrimoines`
2022-07-10 20:50:35 [INFO] - Thread 14 dumping db information for `yseo`
2022-07-10 20:50:35 [INFO] - Thread 6 dumping db information for `yvesgicquelconse`
2022-07-10 20:50:35 [INFO] - Thread 16 dumping db information for `yzial`
2022-07-10 20:50:35 [INFO] - Thread 10 dumping schema create for `12toitplace`
2022-07-10 20:50:35 [INFO] - Thread 10 dumping schema for `12toitplace`.`actions`
2022-07-10 20:50:35 [INFO] - Thread 10 dumping data for `12toitplace`.`actions` | Remaining jobs: 2469550
2022-07-10 20:50:35 [INFO] - Thread 10 dumping schema for `12toitplace`.`actions_groupes`
2022-07-10 20:50:35 [INFO] - Thread 10 dumping schema create for `7actifs`
 
**How to repeat**
Dump a MYSQL (not mariadb) database with a large number of innodb table and a few MYISAM table (not MGR_MYISAM).

**Environment (please complete the following information):**
 - OS version: Debian 11
 - MyDumper version: 0.9.1 and 0.12.3
 - MYSQL Community 5.7.17

**Comment by davidducos** [comment]
I think that it is a duplicate of #648

**Comment by vnlo77250** [comment]
Hi,

what about the issue of MYISAM engine detection ?

**Comment by vnlo77250** [comment]
If i'm not mistaken,  the following line 

if (ecol != NULL && g_ascii_strcasecmp("MRG_MYISAM",ecol)) {

mean that only tables with MERGE engine are considered as non innodb table and will be queued as such.
what about MYISAM,  ARCHIVE and so on ?

**Comment by hustegg** [comment]
Can we exchange the order of join less-locking threads and unlock tables?
I'm not sure if I missed some detail, innodb tables in FTWRL can be unlocked earlier after non-innodb tables locked read local in less-locking mode, instead of all non-innodb tables being dumped.

```
  if (!no_locks && !trx_consistency_only) {
    g_async_queue_pop(conf.unlock_tables);
    g_message("Non-InnoDB dump complete, unlocking tables");
    mysql_query(conn, "UNLOCK TABLES /* FTWRL */");
    g_message("Releasing DDL lock");
    if (release_binlog_function != NULL){
      g_message("Releasing binlog lock");
      release_binlog_function(second_conn);
    }
  }

  if (less_locking) {
    g_message("Waiting less locking jobs to complete");
    for (n = num_threads; n < num_threads * 2; n++) {
      g_thread_join(threads[n]);
    }
    g_async_queue_unref(conf.queue_less_locking);
    conf.queue_less_locking=NULL;
  }
```

**Comment by davidducos** [comment]
Yes, I also noticed that, there is something out of order that I was not able to fix yet. Take into account that less-locking was a solution to a different problem that we need to start using to solve newer issues. This is another thing that I will working for next release on #648

**Comment by davidducos** [comment]
This should be fixed by #847
